contracts/utils/CommitteeUpgradeable.sol

```diff
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "../interfaces/IBridgeCommittee.sol";
+import "./MessageVerifier.sol";
+
+abstract contract CommitteeUpgradeable is
+    UUPSUpgradeable,
+    MessageVerifier,
+    ReentrancyGuardUpgradeable
+{
+    bool private _upgradeAuthorized;
+
+    function __CommitteeUpgradeable_init(address _committee) internal onlyInitializing {
+        __ReentrancyGuard_init();
+        __MessageVerifier_init(_committee);
+        committee = IBridgeCommittee(_committee);
+    }
+
+    function _authorizeUpgrade(address) internal view override {
+        require(_upgradeAuthorized, "SuiBridge: Unauthorized upgrade");
+    }
+
+    function upgradeWithSignatures(bytes[] memory signatures, BridgeMessage.Message memory message)
+        external
+        nonReentrant
+        verifySignatures(message, signatures, BridgeMessage.BRIDGE_UPGRADE)
+    {
+        // decode the upgrade payload
+        (address implementationAddress, bytes memory callData) =
+            BridgeMessage.decodeUpgradePayload(message.payload);
+
+        // authorize upgrade
+        _upgradeAuthorized = true;
+        // upgrade contract
+        upgradeToAndCall(implementationAddress, callData);
```

Comment:
did you have a chance to check if this is reentrance free?

Reply:
upgradeToAndCall is not itself nonReentrant but all "WithSignature" (ex. `upgradeWithSignatures`) functions should be.

Reply:
so the theory is as long as the callData (which is signed by validators) is not reentrant-able, this is safe?

Reply:
correct